### PR TITLE
All Code

### PR DESCRIPTION
**Pull request by forsakenneon**

### CODE REVIEW

File: Mic1/src/main/java/com/m1/db/DBContext.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+						CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build())))
+				.getCollection(collectionName, collectionType);
+	}
+
+	public static <T> MongoCursor<Document> fetchCollectionCursor(String dbName, String collectionName,
+			Class<T> collectionType) {
+		return MongoClients.create(MONGO_URI).getDatabase(dbName)
+				.withCodecRegistry(CodecRegistries.fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
+						CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build())))
+				.getCollection(collectionName).find().iterator();
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/DBContext.java:28]: use fetchCollection(...).find().iterator()

File: Mic1/src/main/java/com/m1/db/repository/UserRepository.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import com.m1.db.entity.User;
+import com.m1.utils.JsonUtil;
+import com.mongodb.BasicDBObject;
+import com.mongodb.DBObject;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.Updates;
+
+public class UserRepository {
+	
+	public static String getFirstName(String id) throws Exception {
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/repository/UserRepository.java:18]: Repository must be injected in service. This don't use static methods.

File: Mic1/src/main/java/com/m1/db/repository/UserRepository.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		MongoCollection<Document> collection = DBContext.fetchCollection("data", "Users", Document.class);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/repository/UserRepository.java:19]: create field in repository for Collection

File: Mic1/src/main/java/com/m1/db/repository/UserRepository.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		Document myDoc = collection.find(eq("_id", id)).first();
+		return JsonUtil.fromJsontoUser(myDoc.toJson()).getFirstName();
+	}
+
+	public static void addFirstName(User user, String id) {
+		try {
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/repository/UserRepository.java:25]: handle exceptions in service. Use `methodName() throws Exception` to throw it out

File: Mic1/src/main/java/com/m1/db/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+			MongoCollection<Document> collection = DBContext.fetchCollection("data", "Users", Document.class);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/repository/UserRepository.java:26]: same

File: Mic1/src/main/java/com/m1/db/repository/UserRepository.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+			Document document = new Document("_id", id);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/repository/UserRepository.java:27]: use var

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.stereotype.Service;
+
+import com.m1.db.entity.User;
+import com.m1.db.repository.UserRepository;
+import com.m1.feign.FeignService;
+import com.m1.user.utils.UserUtils;
+import com.m1.utils.JsonUtil;
+
+@Service
+@Component
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:17]: Remove component. Service contain it.

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+public class UserService {
+	
+	@Autowired
+	private FeignService feign;
+	
+	public User getUser(String id) {
+		User user = new User();
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:24]: User user = null;

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import com.m1.feign.FeignService;
+import com.m1.user.utils.UserUtils;
+import com.m1.utils.JsonUtil;
+
+@Service
+@Component
+public class UserService {
+	
+	@Autowired
+	private FeignService feign;
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:21]: rename feignService

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	
+	public User getUser(String id) {
+		User user = new User();
+		try {
+			user = feign.GetMidWithLastName(id);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:26]: feignService.getMiddleAndLastName

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+			user.setFirstName(UserRepository.getFirstName(id));
+		} catch (Exception e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		return user;
+	}
+	
+	public void postUser(String jsonUser) {
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:35]: createUser*
Return User

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+		String id = UserUtils.generateId();
+		User user;
+		try {
+			user = JsonUtil.fromJsontoUser(jsonUser);
+			UserRepository.addFirstName(user, id);
+			feign.PostMidName(user, id);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:41]: User user = feignService.createMiddleAndLastName

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+		} catch (Exception e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		
+	}
+	
+	public void putUser(String jsonUser, String id) {
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:49]: updateUser
return User

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+		User user;
+		try {
+			user = JsonUtil.fromJsontoUser(jsonUser);
+			UserRepository.updateFirstName(user, id);
+			feign.PutMidName(user, id);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:54]: feignService.updateMiddleAndLastName

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+		} catch (Exception e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:57]: forget to log error

File: Mic1/src/main/java/com/m1/service/UserService.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+		}
+		
+	}
+	
+	public void deleteUser(String id){
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/service/UserService.java:62]: return String "Deleted!"

File: Mic1/src/main/java/com/m1/db/repository/UserRepository.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+			document.put("firstName", user.getFirstName());
+			collection.insertOne(document);
+
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	public static void updateFirstName(User user, String id) throws Exception {
+		MongoCollection<User> collection = DBContext.fetchCollection("data", "Users", User.class);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/repository/UserRepository.java:37]: use field in repository

File: Mic1/src/main/java/com/m1/db/repository/UserRepository.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+		collection.updateOne(new Document("_id", id), Updates.set("firstName", user.getFirstName()));
+	}
+
+	public static void deleteFirstName(String id) {
+		MongoCollection<Document> collection = DBContext.fetchCollection("data", "Users", Document.class);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/db/repository/UserRepository.java:42]: same

File: Mic1/src/main/java/com/m1/feign/FeignService.java [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+
+import com.m1.db.entity.User;
+
+@FeignClient(name = "m1", url = "http://localhost:8092")
+public interface FeignService {
+
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/feign/FeignService.java:15]: getAll

File: Mic1/src/main/java/com/m1/feign/FeignService.java [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+import com.m1.db.entity.User;
+
+@FeignClient(name = "m1", url = "http://localhost:8092")
+public interface FeignService {
+	
+	@GetMapping(value = "/user/get/{id}")
+	User GetMidWithLastName(@PathVariable String id);
+	
+	@PostMapping(value = "/user/add/{id}")
+	String PostMidName(@RequestBody User user, @PathVariable String id);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/feign/FeignService.java:20]: return User

File: Mic1/src/main/java/com/m1/feign/FeignService.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+	@PutMapping(value = "/user/update/{id}")
+	String PutMidName(@RequestBody User user, @PathVariable String id);
```
Comment (KorchevskiyBohdan) [inline Mic1/src/main/java/com/m1/feign/FeignService.java:23]: return User

File: Mic1/src/main/java/com/m1/service/UserService.java (hunk 8 above)
Comment (forsakenneon) [reply Mic1/src/main/java/com/m1/service/UserService.java:24]: If my User = null, than it thows Exception.

File: Mic2/src/main/java/com/m2/api/controller/UserRestController.java [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+	
+	@PostMapping("/add/{id}")
+	@ResponseStatus(HttpStatus.CREATED)
+	public void postUserToMic3(@RequestBody User user, @PathVariable String id) {
+		userService.postUser(user, id);
+	}
+
+	@PutMapping("/update/{id}")
+	public void putUser(@RequestBody User user, @PathVariable String id) {
+		userService.putUser(user, id);
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/api/controller/UserRestController.java:46]: userService.update(user)

File: Mic2/src/main/java/com/m2/api/controller/UserRestController.java [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	@DeleteMapping("/delete/{id}")
+	public void deleteUser(@PathVariable String id) {
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/api/controller/UserRestController.java:50]: delete

File: Mic2/src/main/java/com/m2/api/controller/UserRestController.java [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+		userService.deleteUser(id);
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/api/controller/UserRestController.java:51]: delete

File: Mic2/src/main/java/com/m2/db/DBContext.java [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+			Class<T> collectionType) {
+		return MongoClients.create(MONGO_URI).getDatabase(dbName)
+				.withCodecRegistry(CodecRegistries.fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
+						CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build())))
+				.getCollection(collectionName, collectionType);
+	}
+
+	public static <T> MongoCursor<Document> fetchCollectionCursor(String dbName, String collectionName,
+			Class<T> collectionType) {
+		return MongoClients.create(MONGO_URI).getDatabase(dbName)
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/db/DBContext.java:25]: previous comment

File: Mic2/src/main/java/com/m2/db/repository/UserRepository.java [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+import com.mongodb.DBObject;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.model.Updates;
+
+@Repository
+public class UserRepository {
+	
+	public ArrayList<String> getAllMiddleNames() throws Exception { 
+		MongoCursor<Document> cursor = DBContext.fetchCollectionCursor("data", "Users", Document.class);
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/db/repository/UserRepository.java:24]: its a field

File: Mic2/src/main/java/com/m2/db/repository/UserRepository.java [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+				middleNames.add(JsonUtil.fromJsontoUser(cursor.next().toJson()).getMiddleName());
+			}
+		} finally {
+			cursor.close();
+		}
+		return middleNames;
+	}
+	
+	public static String getMiddleName(String id) throws Exception {
+		MongoCollection<Document> collection = DBContext.fetchCollection("data", "Users", Document.class);
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/db/repository/UserRepository.java:37]: same

File: Mic2/src/main/java/com/m2/db/repository/UserRepository.java [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+
+		if (found != null) {
+			Bson updatevalue = new Document("middleName", user.getMiddleName());
+			Bson updateoperation = new Document("$set", updatevalue);
+			collection.updateOne(found, updateoperation);
+		}
+	}
+
+	public static void updateMiddleName(User user, String id) throws Exception  {
+		MongoCollection<User> collection = DBContext.fetchCollection("data", "Users", User.class);
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/db/repository/UserRepository.java:54]: same

File: Mic2/src/main/java/com/m2/service/UserService.java [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+	private UserRepository userRepository;
+	
+	public ArrayList<User> getAllMiddleAndLastNames() {
+		var middleAndLastNames = new ArrayList<User>();
+		try {
+			var middleNames = userRepository.getAllMiddleNames();
+			var lastNames = webClientService.getAllLastNames();
+			for(int i = 0; i < middleNames.size(); i++) {
+				var user = new User();
+				user.setMiddleName(middleNames.get(i));
```
Comment (KorchevskiyBohdan) [inline Mic2/src/main/java/com/m2/service/UserService.java:31]: constructor

File: Mic3/src/main/java/com/m3/db/DBContext.java [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+		return MongoClients.create(MONGO_URI).getDatabase(dbName)
+				.withCodecRegistry(CodecRegistries.fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
+						CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build())))
+				.getCollection(collectionName, collectionType);
+	}
+
+	public static <T> MongoCursor<Document> fetchCollectionCursor(String dbName, String collectionName,
+			Class<T> collectionType) {
+		return MongoClients.create(MONGO_URI).getDatabase(dbName)
+				.withCodecRegistry(CodecRegistries.fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
```
Comment (KorchevskiyBohdan) [inline Mic3/src/main/java/com/m3/db/DBContext.java:26]: same

File: docker-compose.yml [hunk 31]
```diff
@@ -0,0 +1,40 @@
...
+     network_mode: host
+     depends_on:
+       - MIC1
+
+  MIC3:
+     image: microservice3
+     build:
+       context: ./Mic3
+       dockerfile: Dockerfile
+     ports:
```
Comment (KorchevskiyBohdan) [inline docker-compose.yml:36]: close ports

File: docker-compose.yml [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+       - 8091:8091
+     network_mode: host
+     depends_on:
+       - mongodb
+
+  MIC2:
+     image: microservice2
+     build:
+       context: ./Mic2
+       dockerfile: Dockerfile
```
Comment (KorchevskiyBohdan) [inline docker-compose.yml:24]: close ports

**Pull request merged**